online-shop/src/app/edit-product/edit-product.component.ts

```diff
@@ -0,0 +1,114 @@
+import { Component, OnInit } from '@angular/core';
+import {FormBuilder, FormControl, FormGroup, Validators} from "@angular/forms";
+import {ActivatedRoute} from "@angular/router";
+import {Location} from "@angular/common";
+import {ProductsService} from "../products.service";
+import {tap, first} from "rxjs/operators";
+
+@Component({
+  selector: 'app-edit-product',
+  templateUrl: './edit-product.component.html',
+  styleUrls: ['./edit-product.component.scss']
+})
+export class EditProductComponent implements OnInit {
+
+  product: any;
+
+  editProductForm: FormGroup = new FormGroup({
+    name: new FormControl('', [Validators.required, Validators.pattern("(.|\\s)*\\S(.|\\s)*")]),
+    category: new FormControl('', [Validators.required, Validators.pattern("(.|\\s)*\\S(.|\\s)*")]),
+    image: new FormControl('', [Validators.required, Validators.pattern("(.|\\s)*\\S(.|\\s)*")]),
+    price: new FormControl('', [Validators.required, Validators.pattern("^[0-9]*$")]),
+    description: new FormControl('', [Validators.required, Validators.pattern("(.|\\s)*\\S(.|\\s)*")])
+  })
+
+  //form state
+  loading = false;
+  success = false;
+
+  constructor(private fb: FormBuilder,
+              private productsService: ProductsService,
+              private route: ActivatedRoute,
+              private location: Location) { }
+
+  ngOnInit(): void {
+    this.getProduct()
+
+    this.editProductForm.valueChanges.subscribe(console.log)
+
+  }
+
+  getProduct() {
+    const id = Number(this.route.snapshot.paramMap.get('id'));
+    this.productsService.getProductById(id)
+      .subscribe(product => {
```

Comment:
**suggestion:** take care to avoid memory leaks in your code by unsubscribing from observables. https://www.thisdot.co/blog/best-practices-for-managing-rxjs-subscriptions/